<a href="https://colab.research.google.com/github/HarshESC/Deep-Learning/blob/main/Tensor_Kera_HW.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import warnings
warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

In [ ]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

input_dim = 784  # 28*28
output_dim = nb_classes = 10
batch_size = 128
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

11501568/11490434 [==============================] - 0s 0us/step


In [ ]:
Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

## 1. In this task, you'll build an ANN and train and test it using the MNIST data. This ANN should consist of two hidden and one output layer. All hidden layers should be dense. The neuron sizes of the first layer and the second layer should be 32 and 16 respectively. Train this model 20 epochs and compare your train and test set performance with the example in the checkpoint. Is there any difference? If so, why?


In [ ]:
model = Sequential()
# our first dense layer
model.add(Dense(32, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(16, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

In [ ]:
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

In [ ]:
# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 2s 2ms/step - loss: 1.7141 - accuracy: 0.4305
Epoch 2/20
469/469 [==============================] - 1s 2ms/step - loss: 0.7869 - accuracy: 0.8031
Epoch 3/20
469/469 [==============================] - 1s 2ms/step - loss: 0.5084 - accuracy: 0.8631
Epoch 4/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4165 - accuracy: 0.8863
Epoch 5/20
469/469 [==============================] - 1s 2ms/step - loss: 0.3699 - accuracy: 0.8977
Epoch 6/20
469/469 [==============================] - 1s 2ms/step - loss: 0.3400 - accuracy: 0.9049
Epoch 7/20
469/469 [==============================] - 1s 2ms/step - loss: 0.3187 - accuracy: 0.9104
Epoch 8/20
469/469 [==============================] - 1s 2ms/step - loss: 0.3024 - accuracy: 0.9141
Epoch 9/20
469/469 [==============================] - 1s 2ms/step - loss: 0.2887 - accuracy: 0.9175
Epoch 10/20
469/469 [==============================] - 1s 2ms/step - loss: 0.2770 - accuracy: 0.9203

In [ ]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.19878371059894562
Test accuracy: 0.9416000247001648


The model achieved 94% accuracy in both the training and the test set. This level is lower than that of the example in the checkpoint. Since the layers include less number of neurons in this model, the model is simpler than the one of the example. This resulted in a lower performance. It seems, MNIST data requires a more complicated model than this one.

## 2. You'll also build an ANN in this task. This time, this ANN should have 5 hidden layers and 1 output layer. All the layers should be dense. The neuron numbers for the hidden layers should be 1024, 512, 256, 128 and 64. Train this model 20 epochs and test it using the same data from the previous task and compare your results. Is there any difference? If so, why?

In [ ]:
model = Sequential()
# our first dense layer
model.add(Dense(1024, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(512, activation="relu"))
# our third dense layer
model.add(Dense(256, activation="relu"))
# our fourth dense layer
model.add(Dense(128, activation="relu"))
# our fifth dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

In [ ]:
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

In [ ]:
# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 13s 26ms/step - loss: 1.2224 - accuracy: 0.6916
Epoch 2/20
469/469 [==============================] - 12s 27ms/step - loss: 0.3674 - accuracy: 0.8984
Epoch 3/20
469/469 [==============================] - 12s 26ms/step - loss: 0.2764 - accuracy: 0.9203
Epoch 4/20
469/469 [==============================] - 13s 27ms/step - loss: 0.2307 - accuracy: 0.9327
Epoch 5/20
469/469 [==============================] - 13s 27ms/step - loss: 0.2013 - accuracy: 0.9412
Epoch 6/20
469/469 [==============================] - 12s 27ms/step - loss: 0.1797 - accuracy: 0.9481
Epoch 7/20
469/469 [==============================] - 13s 27ms/step - loss: 0.1612 - accuracy: 0.9532
Epoch 8/20
469/469 [==============================] - 13s 27ms/step - loss: 0.1461 - accuracy: 0.9578
Epoch 9/20
469/469 [==============================] - 12s 27ms/step - loss: 0.1330 - accuracy: 0.9614
Epoch 10/20
469/469 [==============================] - 13s 27ms/step - loss: 0.122

In [ ]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.09635999798774719
Test accuracy: 0.9704999923706055


MUCH better.  97% accuracy is higher than the example one of 95%